File: camerax_android/lib/src/legacy/video/quality_selector.dart
import 'package:camerax_android/src/legacy/camerax.g.dart' as $native;
import 'package:camerax_android/src/legacy/common.dart';
import 'package:camerax_android/src/legacy/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;

import 'fallback_strategy.dart';
import 'quality.dart';

final class QualitySelector extends $interface.QualitySelector {
  final $native.QualitySelector obj;

  QualitySelector.$native(this.obj) : super.impl();

  factory QualitySelector.from(
    $interface.Quality quality, {
    $interface.FallbackStrategy? fallbackStrategy,
  }) {
    if (fallbackStrategy is! FallbackStrategy?) {
      throw TypeError();
    }
    final obj = $native.QualitySelector.from(
      quality: quality.obj,
      fallbackStrategy: fallbackStrategy?.obj,
    );
    return QualitySelector.$native(obj);
  }

  factory QualitySelector.fromOrderedList(
    List<$interface.Quality> qualities, {
    $interface.FallbackStrategy? fallbackStrategy,
  }) {
    if (fallbackStrategy is! FallbackStrategy?) {
      throw TypeError();
    }
    final obj = $native.QualitySelector.fromOrderedList(
      qualities: qualities.map((quality) => quality.obj).toList(),
      fallbackStrategy: fallbackStrategy?.obj,
    );
    return QualitySelector.$native(obj);
  }

  static Future<$interface.Size?> getResolution(
      $interface.CameraInfo cameraInfo, $interface.Quality quality) async {
    if (cameraInfo is! CameraInfo) {
      throw TypeError();
    }
    final obj = await $native.QualitySelector.getResolution(
        cameraInfo.obj, quality.obj);
    return obj?.args;
  }
}

File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/legacy/video/QualitySelector.kt
package dev.hebei.camerax_android.legacy.video

import dev.hebei.camerax_android.legacy.CameraXRegistrar
import dev.hebei.camerax_android.legacy.PigeonApiQualitySelector

class QualitySelector(registrar: CameraXRegistrar) : PigeonApiQualitySelector(registrar) {
    override fun from(
        quality: androidx.camera.video.Quality, fallbackStrategy: androidx.camera.video.FallbackStrategy?
    ): androidx.camera.video.QualitySelector {
        return if (fallbackStrategy == null) androidx.camera.video.QualitySelector.from(quality)
        else androidx.camera.video.QualitySelector.from(quality, fallbackStrategy)
    }

    override fun fromOrderedList(
        qualities: List<androidx.camera.video.Quality>, fallbackStrategy: androidx.camera.video.FallbackStrategy?
    ): androidx.camera.video.QualitySelector {
        return if (fallbackStrategy == null) androidx.camera.video.QualitySelector.fromOrderedList(qualities)
        else androidx.camera.video.QualitySelector.fromOrderedList(qualities, fallbackStrategy)
    }

    override fun getResolution(
        cameraInfo: androidx.camera.core.CameraInfo, quality: androidx.camera.video.Quality
    ): android.util.Size? {
        return androidx.camera.video.QualitySelector.getResolution(cameraInfo, quality)
    }
}